### PR TITLE
QIR generation bug fixes

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR fixes https://github.com/microsoft/qsharp-compiler/issues/1152, https://github.com/microsoft/qsharp-compiler/issues/1151, https://github.com/microsoft/qsharp-compiler/issues/1086, https://github.com/microsoft/qsharp-compiler/issues/1159, https://github.com/microsoft/qsharp-compiler/issues/1160, https://github.com/microsoft/qsharp-compiler/issues/1161, https://github.com/microsoft/qsharp-compiler/issues/1162, https://github.com/microsoft/qsharp-compiler/issues/1164, and https://github.com/microsoft/qsharp-compiler/issues/1096.

This PR also removes the use of __quantum__rt__pauli_to_string, and adjust the signature of __quantum__rt__array_slice_1d to match the runtime implementation (which deviates from spec - this will be remedied shortly).

### CODE REVIEW

File: src/QsCompiler/QirGeneration/Interop.cs [hunk 1]
```diff
@@ -539,13 +539,8 @@ private IrFunction GenerateEntryPoint(string entryPointName, ArgumentTuple argum
 
                 // print the return value
                 var message = this.sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.Message);
-                var outputStr = QirExpressionKindTransformation.CreateStringLiteral(
-                    this.sharedState,
-                    result.QSharpType.Resolution.IsString ? "\"{0}\"" : "{0}",
-                    result);
+                var outputStr = QirExpressionKindTransformation.CreateStringLiteral(this.sharedState, "{0}", result);
```
Comment (cesarzc) [inline src/QsCompiler/QirGeneration/Interop.cs:542]: Before this change, if the type was string, it was wrapped around additional quotation marks. Why is it no longer the case?

File: src/QsCompiler/QirGeneration/QIR/RuntimeLibrary.cs [hunk 2]
```diff
@@ -27,13 +27,11 @@ public static class RuntimeLibrary
...
         // to-string
         public const string BigIntToString = "bigint_to_string";
-        public const string BoolToString = "bool_to_string";
         public const string DoubleToString = "double_to_string";
         public const string IntToString = "int_to_string";
-        public const string PauliToString = "pauli_to_string";
         public const string QubitToString = "qubit_to_string";
-        public const string RangeToString = "range_to_string";
-        public const string ResultToString = "result_to_string";
+        public const string RangeToString = "range_to_string"; // todo: eliminate
```
Comment (cesarzc) [inline src/QsCompiler/QirGeneration/QIR/RuntimeLibrary.cs:33]: Do these `todo`s need to be addressed before merging this case? If they will be kept for this change, it would be good to comment on when/why the strings need to be eliminated.

File: src/QsCompiler/QirGeneration/Interop.cs (hunk 1 above)
Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Interop.cs:542]: With the latest change, interpolation arguments of type string are wrapped inside quotation marks, hence there is no need to additionally wrap it once more.

File: src/QsCompiler/QirGeneration/QIR/RuntimeLibrary.cs (hunk 2 above)
Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/QIR/RuntimeLibrary.cs:33]: No, this is a todo related to a spec update - or rather, it is possible (and likely reasonable) to not relay on the runtime for this functionality, but also lower priority compared to the general fixes in this PR.

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs [hunk 3]
```diff
@@ -262,55 +267,91 @@ void StoreElement(PointerValue pointer, IValue value, string? fromLocalId, bool
...
+
                     // do not increase the ref count here - we will increase the ref count of all new items at the end
                     IValue newItemValue = sharedState.EvaluateSubexpression(updated);
                     var index = sharedState.EvaluateSubexpression(accEx);
-                    UpdateElement(_ => newItemValue, index.Value);
+                    UpdateElement(() => newItemValue, array.GetArrayElementPointer(index.Value));
                 }
                 else if (accEx.ResolvedType.Resolution.IsRange)
                 {
+                    // fixme: if fromLocalId equals the lhs then we need to make a deep copy of updated ...
```
Comment (cesarzc) [inline src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:306]: Should this fixme be tracked as an issue?

Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs]: This comment is just a remainder of my notes, and I have revised the comment here to be a description of what is done.

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs [hunk 4]
```diff
@@ -584,16 +658,21 @@ Value ArrayToString(ArrayValue array)
                 var ty = evaluated.QSharpType.Resolution;
                 if (ty.IsString)
                 {
-                    UpdateStringRefCount(evaluated.Value, 1);
-                    return evaluated.Value;
+                    Value quote = CreateConstantString("\"");
```
Comment (cesarzc) [inline src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:661]: Nit: Is there a reason why `quote` was created using the `Value` type instead of `var`? I see that the `comma` variable a few lines above uses the same pattern.

Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:661]: I think this might have been left from an earlier stage of the code where the explicit Value was needed for the nullcheck to be happy. I'll change it to var.

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs [hunk 5]
```diff
@@ -605,7 +684,29 @@ Value ArrayToString(ArrayValue array)
                 }
                 else if (ty.IsPauli)
                 {
-                    return SimpleToString(RuntimeLibrary.PauliToString);
+                    Value LoadPauli(QsPauli pauli)
```
Comment (cesarzc) [inline src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:687]: It seems like a rather complicated change to avoid calling `PauliToString`. Why the `Pauli` type needs to be handled this way?

Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs:687]: There are a couple of general considerations:
1) Generally, I think it is good to trim down what the runtime needs to implement to keep things simple for backends. 
2) If the runtime implements it, then the compiled code will break down into something similar, so there is no advantage or disadvantage really one way or the other after all is compiled.
3) The disadvantage of implementing it in the compiler is a little bit more code complexity compared to implementing it in the runtime (the code for us to maintain, not the code that a Q# project compiles into).

What the reason was for the immediate change is that with the current implementation of the runtime, there can be issues with bool and Pauli, since these don't match the bitsize used in the IR. This is only relevant in edge cases, and the proper solution is to revise the API to avoid such incompatibilities. We will do this as part of moving to a C-API. In the meantime, avoiding to use things we know can cause issues (for edge cases, most of the time the runtime functions work fine as is) is a mitigation.

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/ExecutionTests.qs [hunk 6]
```diff
@@ -1048,11 +1614,59 @@
...
+    operation ReturnsString() : String {
+        return "Success!";
+    }
+
     @EntryPoint()
     operation RunExample() : Unit {
 
         PrintSection(1, "");
-
+
```
Comment (cesarzc) [inline src/QsCompiler/Tests.Compiler/TestCases/QirTests/ExecutionTests.qs:1665]: Nit: it seems like the editor might have added a few space characters.

**Review by cesarzc — APPROVED** [review]

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/ExecutionTests.qs (hunk 6 above)
Comment (bettinaheim) [reply src/QsCompiler/Tests.Compiler/TestCases/QirTests/ExecutionTests.qs:1665]: I'll wait to have a formatter to remove these automatically. :) They occur when editing multiple lines, e.g. commenting them out.
